functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,235 @@
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+)
+
+// the similar struct esxits in resid.GVK, but there is no function to create an GVK struct without using kyaml
+type FieldSpec struct {
+	Group              string `json:"group,omitempty" yaml:"group,omitempty"`
+	Version            string `json:"version,omitempty" yaml:"version,omitempty"`
+	Kind               string `json:"kind,omitempty" yaml:"kind,omitempty"`
+	Path               string `json:"path,omitempty" yaml:"path,omitempty"` // seperated by /
+	CreateIfNotPresent bool   `json:"create,omitempty" yaml:"create,omitempty"`
+	//TODO: should support user configurable field for labels
+}
+
+func SetLabels(rl *fn.ResourceList) (bool, error) {
+	transformer := LabelTransformer{}
+	if err := transformer.Config(rl.FunctionConfig); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, err
+	}
+	if err := transformer.Transform(rl.Items); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, err
+	}
+
+	rl.Results = append(rl.Results, transformer.Results...)
+	return true, nil
+}
+
+type LabelTransformer struct {
+	// Desired labels
+	NewLabels map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
+	// FieldSpecs storing default label fields
+	FieldSpecs []FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+	// Results is used to track labels that have been applied
+	Results fn.Results
+}
+
+// Config parse the functionConfig kubeObject to the fields in the LabelTransformer
+func (p *LabelTransformer) Config(functionConfig *fn.KubeObject) error {
+	// parse labels to NewLabels
+	switch {
+	case functionConfig.IsEmpty():
+		return fmt.Errorf("failed to configure function: `functionConfig` must be either a `ConfigMap` or `SetLabels`")
+	case functionConfig.IsGVK("", "v1", "ConfigMap"):
+		p.NewLabels = functionConfig.NestedStringMapOrDie("data")
+	case functionConfig.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind):
+		if _, exist, err := functionConfig.NestedSlice(fnDeprecateField); exist || err != nil {
+			return fmt.Errorf("`additionalLabelFields` has been deprecated")
+		}
+		p.NewLabels = functionConfig.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty, required valid `labels` field")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			functionConfig.GetKind(), functionConfig.GetAPIVersion(), fnConfigKind)
+	}
+	return nil
+}
+
+// Transform updates the labels in the right path using configured logic
+func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+
+	for _, o := range objects {
+		var path = "metadata/labels"
+		err := updateLabels(o, path, p.NewLabels, true)
+		if err != nil {
+			return err
+		}
+		p.LogResult(o, path)
+		// check all default path to update labels
+		switch {
```

Comment:
I believe you can simplify the logic below by using the config of `CommonLabelFieldSpecs` and `fieldSpec`.
Currently it seems you are hardcoding it below.

Reply:
I thought it is best not to use json or yaml unmarsh to read in the **CommonLabelFieldSpecs**, and I should use **IsGVK**, which is a bit hard-coding as I write here. Is this what you expect or I misunderstood the point? @yuwenma

Reply:
I think we want to avoid `path = "metadata/labels"` as well.

Reply:
@mengqiy and I have an offline discussion about the code style here.
Something you can consider.

```go
type FieldPath []string
type GVK []string
// for fieldpath that has simple stringMap field.   
stringMapLabels := map[fn.ResourceIdentifier]FieldPath{
        GVK{"", "v1", "Service"}: FieldPath{"spec", "selectors"} 
      .........
}

for _, o := range objects {
          o.SetLabels()... // or SetLabel
            // for pure stringMap fieldspec
           setLabelsInSpecs(o)
           // handle special cases
           if o.IsGVK("apps", "", "StatefulSet") {
                  for _, vctObj:=  range o.GetMap("spec").GetSlice("volumeClaimTemplates")  {
                        setVolumeClaimTemplatesLabel(vctObj)
                    }            
           }
}

func setLabelsInSpecs(o fn.KubeObject) {
           for gvk, paths := range stringMapLabels { 
                     if o.IsGVK(gvk...) {
                             o.SetNestedStringMap(newLabels, paths...) 
                     }
            }
}
```